Sources/SwiftParser/Recovery.swift

```diff
@@ -123,13 +123,29 @@ extension Parser.Lookahead {
       if currentTokenPrecedence >= recoveryPrecedence {
         break
       }
-      self.consumeAnyToken()
       if let closingDelimiter = currentTokenPrecedence.closingTokenKind {
         let closingDelimiterSpec = TokenSpec(closingDelimiter)
+        let canSkipSingleAtSameLine = self.withLookahead { lookahead in
```

Comment:
I don’t quite understand why this works. `canSkipSingleAtSameLine` checks if we have the closing delimiter on the  same line and allows skipping over tokens with a higher precedence. For example, when `closingDelimiter` was `)` and there was a `}` in between, it would skip over that. But then `skipSingle` uses a different logic to essentially discover the closing token again (which doesn’t match for the `)` and `}` I described). 

If the intended behavior is to skip until the closing token on the same line, skipping over any higher precedence tokens (which I think makes sense, it produces good results in `testRecovery105`), I think we should count the number of tokens consumed in the nested lookahead and consume them, not calling into `skipSingle`.

Reply:
TBH I came up with this solution after trying just `skipSingle`, which improved `testRecovery105' but broke other test, so I added this lookahead to prevent too much skipping. Your suggestion does make more sense. I will try it, thanks!